Store days separately in Intervals

jkosh44 opened the pull request:
Previously Intervals stored days, hours, minutes, and seconds all in a
single field. They were all converted to nanoseconds and stored in an
i128. This is different than PostgreSQL, which stores days in a
separate field. This commit changes the internal Interval
representation to match PostgreSQL by storing days separately. It also
changed the i128 nanosecond to a i64 microsecond to match PostgreSQL.

This results in some new behavior around comparing Intervals. Equality
does not convert between months, days, and microseconds. So
INTERVAL '24 hours' does not equal INTERVAL '1 day'. This is a
limitation introduced by dataflow that requires equivalent Datums to
have identical in memory representations.

This commit also makes some small changes to printing and rounding
Intervals to make it closer to the PostgreSQL behavior.

Fixes MaterializeInc/database-issues#3151

### Motivation

This PR refactors existing code.

### Tips for reviewer

Pay close attention to the `add_field` method in `interval.rs`. That's where the most complicated conversions are and in my opinion the most likely place for bugs.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### Failing Tests
- [x] table_func.slt hangs on `SELECT * FROM generate_series('2021-01-01 03:00:00'::TIMESTAMP, '2021-01-03 00:00:00'::TIMESTAMP, '-1 day') ORDER BY 1`
- [x] Fix date_bin function
- [x] `explain-timers.td` fails
- [x] `check-from-any_version-interval-literals.td` fails

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - **Breaking change.** Store days separately in [`interval`](/sql/types/interval).
  Unlike the previous version, Hours are never converted to days. This
  means that an Interval of 24 hours will not be equal to an Interval
  of 1 day, you cannot subtract hours from days, and when ordering
  Intervals d days > h hours for all d,h {{% gh 10708 %}}.

- **Breaking change.** Print all negative [`interval`](/sql/types/interval)
  units as plural. So `-1 days` will be  printed instead of `-1 day`. This
  matches the behavior of PostgreSQL.

- **Breaking change.** Round microsecond field of
  [`interval`](/sql/types/interval) to 6 places before applying the
  given precision. For example `INTERVAL '1.2345649' SECOND(5)` will be
  rounded to `00:00:01.23457` not `00:00:01.23456`. This matches the
  behavior of PostgreSQL.

jkosh44 commented on src/repr/src/adt/datetime.rs:
```diff
@@ -470,46 +444,46 @@ impl ParsedDateTime {
+        micros: &mut i64,
     ) -> Result<(), String> {
         use DateTimeField::*;
+        fn div_and_round(i: i128, d: i64) -> Option<i64> {
```
It seems strange and imprecise to round the microsecond field every time it's updated, before the precision field is even applied, but I'm pretty sure this is actually the same behavior as postgres: 
https://github.com/postgres/postgres/blob/62cb7427d1e491faf8612a82c2e3711a8cd65422/src/backend/utils/adt/datetime.c#L499-L515

That function is called every time there's a fractional part of some unit. `fsec` there is eventually converted directly to microseconds field and the man pages of `rint` say
> Return the integer nearest X in the direction of the prevailing rounding mode.

Still, I'm curious as to what people think about this.

jkosh44 replied:
The following test does end up testing this behavior and is consistent with postgres, but I think I can definitely add a couple more
```
query T
SELECT INTERVAL '-1.555555555 years 2.555555555 months -3.555555555 days 4.555555555 hours -5.555555555 minutes 6.555555555 seconds';
----
-1 years -4 months +13 days 07:07:53.220829
```

sploiselle replied:
Awesome--as long as this is consistent with PG, despite being unintuitive, seems like the right call

jkosh44 commented on doc/user/content/release-notes.md:
```diff
@@ -107,6 +107,22 @@ These changes are present in [unstable builds](/versions/#unstable-builds) and
+  Unlike the previous version, Hours are never converted to days. This
+  means that an Interval of 24 hours will not be equal to an Interval 
+  of 1 day, you cannot subtract hours from days, and when ordering
+  Intervals d days > h hours for all d,h {{% gh 10708 %}}.
```
I'm still trying to figure out if it's possible to properly order Interval's by converting 24 hrs to 1 day without doing the same thing with equality. The docs for the `PartialOrd` trait say
> The methods of this trait must be consistent with each other and with those of PartialEq in the following sense:
a == b if and only if partial_cmp(a, b) == Some(Equal).
...

https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html

mjibson replied:
Do we have a similar behavior with months, where any number of months is greater than any number of days? Can you add some tests for these to assert our incorrect behavior with a comment linking to some github issue about it?

jkosh44 replied:
The behavior is the same with months, it's just that behavior already existed so I didn't include it in the release notes. I'll add some tests and include a GH link.

jkosh44 commented on src/expr/src/relation/func.rs:
```diff
@@ -962,12 +962,12 @@ fn generate_series_ts(
     conv: fn(NaiveDateTime) -> Datum<'static>,
 ) -> Result<impl Iterator<Item = (Row, Diff)>, EvalError> {
-    if step.months == 0 && step.duration == 0 {
+    if step.months == 0 && step.days == 0 && step.micros == 0 {
```
These need to include days. Probably should combine days and times here, check PG.

jkosh44 commented on src/expr/src/scalar/func.rs:
```diff
@@ -1917,14 +1917,12 @@ where
             "timestamps cannot be binned into intervals containing months or years".to_string(),
         ))
-    } else if stride.duration <= 0 {
+    } else if stride.micros <= 0 {
```
Note to self: These need to include days. Probably should combine days and times here, check PG.

jkosh44 commented on src/repr/src/adt/interval.rs:
```diff
@@ -25,57 +25,65 @@ use crate::adt::numeric::DecimalLike;
+    /// Irrespective of values, `micros` will not be carried over into `days` or
     /// `months`.
-    pub duration: i128,
+    pub micros: i64,
```
Comment/Question: One consequence of this change is that Interval's precision decreases from nanosecond to microsecond Is this something we care about? The only potentially negative thing I've noticed so far is that we use Interval's to print out some of the timing output of `EXPLAIN` queries. This results in the last couple of decimal places being truncated.

sploiselle replied:
PG doesn't return nanoseconds, so I believe this is totally fine. The first implementation of interval in Materialize was _not_ focused on PG interoperability, so had a number of differences; this is likely something from that first iteration that I didn't think to make consistent with PG during my refactor of it.

One thing this did get me to realize, though, is that sub microsecond values do get rounded to microseconds:

```sql
SELECT INTERVAL '01:00:01.00000009';
 interval 
----------
 01:00:01
```

I'm still reviewing, but this looks like something we definitely want to ensure we test/catch.

jkosh44 replied:
Good point, I've added the following tests for this and it seems to be consistent with Postgres as well.
```
query T
SELECT INTERVAL '01:00:01.00000009';
----
01:00:01

query T
SELECT INTERVAL '01:00:01.0000009';
----
01:00:01.000001

query T
SELECT INTERVAL '01:00:01.0000004';
----
01:00:01
```

jkosh44 commented on test/upgrade/check-from-any_version-interval-literals.td:
```diff
@@ -27,7 +27,7 @@
 02:03:00
 
 > SELECT * FROM interval_hour;
-"5 days 03:00:00"
```
Is this OK to change? The new binary format for `Interval`s is definitely not backwards compatible with previous versions. Do we care about that right now? Do I need to write some type of migration? Or is fine for now because persistence isn't fully in place yet?

jkosh44 commented on test/upgrade/check-from-any_version-interval-literals.td:
```diff
@@ -27,7 +27,7 @@
 
 > SELECT * FROM interval_hour;
-"5 days 03:00:00"
+"123:00:00"
```
Is this OK to change? The new binary format for Intervals is definitely not backwards compatible with previous versions. Do we care about that right now? Do I need to write some type of migration? Or is fine for now because persistence isn't fully in place yet?

sploiselle replied:
We'll need version specific tests for this because it won't work w/ any version. You can find the details about what to do here: test/upgrade/README.md but lmk if you want to walk through it

jkosh44 commented:
This PR is now ready for review.

sploiselle commented on src/repr/src/row.proto:
```diff
@@ -106,13 +106,8 @@ message ProtoTimestamp {
-    // of nanos.
-    int64 duration_lo = 2;
-    int64 duration_hi = 3;
+    int32 days = 2;
```
@danhhz Is this OK? Are our persisted values baked in enough yet that we need to write migrations?

danhhz replied:
yup we're still good! we do have the background system tables test, but all that's needed to avoid breaking that is to bump mz_persist::gen::ProtoMeta::ENCODING_VERSION up by 1 and then fix up the resulting unit test breakage (golden_test at least, I don't recall offhand if there are others).

danhhz replied:
Yeah, that TODO is just saying the cleanup _can_ be done now, not that it has to be done on any particular sort of timeline. Definitely go ahead and leave that to us :)

sploiselle reviewed (commented):
What a slog! Let's see what Dan says about the persistence format before moving forward w/ implementing anything else mentioned here.

jkosh44 commented on src/expr/src/scalar/func/impls/interval.rs:
```diff
@@ -28,13 +28,10 @@ sqlfunc!(
         // Negative durations have their HH::MM::SS.NS values subtracted from 1 day.
-        if i.duration < 0 {
-            i = Interval::new(0, 86400, 0)
+        if i.duration_as_chrono() < chrono::Duration::zero() {
```
Since time doesn't contain days, we can just drop the days. I have the following tests in `interval.stl`

```
query T
SELECT INTERVAL '6 days'::time;
----
00:00:00

query T
SELECT INTERVAL '7 months'::time;
----
00:00:00
```

Which matches the behavior of postgres:
```
postgres=# SELECT INTERVAL '6 days'::time;
   time   
----------
 00:00:00
(1 row)

postgres=# SELECT INTERVAL '7 months'::time;
   time   
----------
 00:00:00
(1 row)
```

jkosh44 commented on src/interchange/src/avro/encode.rs:
```diff
@@ -340,12 +340,13 @@ impl<'a> mz_avro::types::ToAvro for TypedDatum<'a> {
                 // so no one is likely to be using it, so we're just using
                 // our own very convenient format.
-                ScalarType::Interval => Value::Fixed(20, {
+                ScalarType::Interval => Value::Fixed(28, {
```
I updated the code, but I'm still a little confused as to where the 24 comes from in the original code. I changed it to 20 to be consistent with lowering it by 4.

sploiselle replied:
24 ensures that the allocation is totally word-aligned (bc the elements are raw bytes), so we want the value to just be 16 here, i.e. should be `x % 8 == 0`.

jkosh44 replied:
Got it, updated accordingly

jkosh44 commented on src/repr/src/adt/datetime.rs:
```diff
@@ -201,49 +201,21 @@ impl DateTimeField {
-    ///
-    /// Panics if called on a non-duration field or a field larger than or equal to a second.
-    pub fn seconds_divider(self) -> i64 {
+    /// Panics if called on a non-duration field.
```
The definition of duration has become a little muddied with this change. Some things require both days and micros while others require just micros. I changed the comment to be a little clearer `Panics if called on a non time/day field.`

jkosh44 commented on src/repr/src/adt/datetime.rs:
```diff
@@ -349,6 +321,8 @@ impl DateTimeFieldValue {
         DateTimeFieldValue { unit, fraction }
     }
+
+    const FRACTION_MULTIPLIER: i64 = 1_000_000_000;
```
How's the new name and comments?
```
/// How much padding is added to the fractional portion to achieve a given precision.
    /// e.g. with the current precision `.5` is represented as `500_000_000`.
    const FRACTIONAL_DIGIT_PRECISION: i64 = 1_000_000_000;
```

danhhz reviewed (commented):
thanks for doing this! when I was writing the storage format, I found it very odd that it worked the way that it did, this is much better.

the row serialization stuff LGTM modulo the ENCODING_VERSION bump I mentioned

mjibson commented on src/expr/src/scalar/func.rs:
```diff
@@ -1932,18 +1932,22 @@ pub fn date_bin<'a, T>(stride: Interval, source: T, origin: T) -> Result<Datum<'
+        ));
+    }
+
+    let stride_ns = match stride.duration_as_chrono().num_nanoseconds() {
```
I'm kind of interested in removing chrono from this. Can we have a num_nanoseconds method directly on interval that doesn't go through chrono? Can you do the math to see if the max/min intervals can always be converted safely to i64 nanos? If not, then it should return a Result/Option. Similarly to the above, the interval type should be in the business of safely converting itself to various forms so that callers don't have to remember to do it safely.

mjibson commented on src/expr/src/scalar/func/impls/interval.rs:
```diff
@@ -28,13 +28,10 @@ sqlfunc!(
         // Negative durations have their HH::MM::SS.NS values subtracted from 1 day.
-        if i.duration < 0 {
-            i = Interval::new(0, 86400, 0)
+        if i.duration_as_chrono() < chrono::Duration::zero() {
```
Can we avoid calls to chrono here?

jkosh44 replied:
Yeah, just removed it.

mjibson commented on src/expr/src/scalar/func/impls/time.rs:
```diff
@@ -30,8 +30,8 @@ sqlfunc!(
-            t.num_seconds_from_midnight() as i64,
-            t.nanosecond().into(),
+            0,
+            (t.num_seconds_from_midnight() * 1_000_000 + t.nanosecond() / 1_000).into(),
```
Lots of unchecked arithmetic here. Needs justification or a checks.

jkosh44 replied:
Just pushed some clean up, with comments justifying why it won't overflow.

mjibson commented on src/pgrepr/src/value/interval.rs:
```diff
@@ -42,22 +42,8 @@ impl ToSql for Interval {
-            days.try_into()
-                .expect("days is bound between i32::MAX and i32::MIN above"),
-        );
+        out.put_i64(self.0.micros);
```
Yesss.

mjibson commented on src/repr/src/adt/interval.rs:
```diff
@@ -107,18 +115,17 @@ impl Interval {
-        let seconds =
-            self.dur_as_secs() as f64 * other + months.fract() * 60.0 * 60.0 * 24.0 * 30.0;
-        if seconds.is_nan() || seconds < i64::MIN as f64 || seconds > i64::MAX as f64 {
+        let days = self.days as f64 * other + months.fract() * 30.0;
```
This function already did all this and you're just changing it but: there's lots of unchecked multiplication, addition, and `as` casting in this function. Can you either check all those or write comments describing why they are guaranteed to be safe?

jkosh44 replied:
I refactored this to remove as much of the `as` as I could. As far as I can tell, the only way to do a lossy conversion between `f64` and `i64`/`i32` is with `as` so I had to leave a couple of them in. If there's a better way please let me know because I couldn't figure it out.

There's no check arithmetic because it's all floating point arithmetic which instead of overflowing/underflowing will go to +inf/-inf. I've added checks for those and returned `None` in those cases.

mjibson commented on src/repr/src/adt/interval.rs:
```diff
@@ -127,23 +134,27 @@ impl Interval {
+        Self::new(months as i32, days as i32, micros as i64).ok()
     }
 
     /// Returns the total number of whole seconds in the `Interval`'s duration.
```
The comment here and below now seem misleading. If I were to read "total" I'd expect months and days to be apart of it. Can you reword?

jkosh44 replied:
I agree, I ended up removing these functions.

mjibson commented on src/repr/src/adt/interval.rs:
```diff
@@ -268,6 +279,7 @@ impl Interval {
     {
         T::from(self.years() * 60 * 60 * 24) * T::from(365.25)
             + T::from(self.months() * 60 * 60 * 24 * 30)
+            + T::from(self.days * 60 * 60 * 24)
```
This can overflow is days is i32::MAX.

jkosh44 replied:
I replaced all the potentially overflowing arithmetic with float arithmetic, which should return +inf/-inf instead of overflowing. I also added tests that use this function with the min/max Intervals

mjibson commented on src/repr/src/adt/interval.rs:
```diff
@@ -295,29 +308,23 @@ impl Interval {
-        // https://github.com/chronotope/chrono/pull/426
-        Duration::seconds(self.dur_as_secs() as i64)
-            + Duration::nanoseconds(self.nanoseconds() as i64)
+        Duration::days(self.days.into()) + Duration::microseconds(self.micros)
```
Months?

sploiselle replied:
Ah, yeah--this is a little goofy. Not sure what the expectation of this function is but it might also want to instead _not_ account for days.

mjibson commented:
What's the minimum amount of chrono we need for this? I think using chrono should be done only when we are forced to because an interval has to interact with some other type that's backed by chrono, but otherwise it should avoid chrono completely.

jkosh44 commented on doc/user/content/release-notes.md:
```diff
@@ -110,6 +110,22 @@ changes that have not yet been documented.
+  Intervals d days > h hours for all d,h {{% gh 10708 %}}.
+
+- **Breaking change.** Print all negative [`interval`](/sql/types/interval)
+  units as plural. So `-1 days` will be  printed instead of `1 day`. This
```
Yes, thank you. Fixed.

jkosh44 commented on src/expr/src/relation/func.rs:
```diff
@@ -961,12 +961,15 @@ fn generate_series_ts(
     conv: fn(NaiveDateTime) -> Datum<'static>,
 ) -> Result<impl Iterator<Item = (Row, Diff)>, EvalError> {
-    if step.months == 0 && step.duration == 0 {
+    let normalized_step = i64::from(step.months) * 30 * 24 * 60 * 60 * 1_000_000
```
I created the `as_microsecond()` method to clean this up. Instead of returning an `Option` or `Result` I just had it return an `i128`. I don't think we want an overflow to cause these checks to fail, since we only really use `normalized step` to check the direction and equality to 0.

mjibson commented on src/repr/src/adt/interval.rs:
```diff
@@ -272,16 +294,57 @@ impl Interval {
+    /// Converts this `Interval`'s duration into `chrono::Duration`.
+    pub fn duration_as_chrono(&self) -> chrono::Duration {
+        use chrono::Duration;
+        Duration::days(self.days.into()) + Duration::microseconds(self.micros)
```
Still scared this doesn't have months in it. Should assert or something.

jkosh44 replied:
I'm just keeping this consistent with the previous implementation which used both days and microseconds, but didn't use months. Honestly not sure right now if it was a bug that the previous implementation didn't use months though.

@sploiselle any thoughts? This is used in the following functions

- `add_date_interval`
- `add_time_interval`
- `add_timestamplike_interval`
- `sub_date_interval`
- `sub_time_interval`
- `timezone_interval_time`
- `timezone_interval_timestamp`
- `timezone_interval_timestamptz`

If this is an existing bug, I'm somewhat inclined to maintain the previous behavior and fix the bug in a followup issue/PR since this PR has grown pretty big. Thought?

sploiselle replied:
Ah, sorry I missed this earlier; this simply doesn't have months in it because month addition was handled by a separate function, e.g. `add_timestamp_months`. I think we can tidy up how all of this works in a separate, follow-on PR. Essentially, we just need to figure out what the semantics are of adding unit-crossing values (e.g. 100 days, 100 months) to these other units. As long as this implementation holds the same semantics as the previous implementation, I think it's fine to move to follow-on work as long as there's bandwidth to do it immediately.

If you plan to go back through CI, @jkosh44, add a `TODO` here.

jkosh44 replied:
Just added a TODO to the function

mjibson commented on src/repr/src/adt/interval.rs:
```diff
@@ -272,16 +294,57 @@ impl Interval {
+        if self.duration_as_chrono() < chrono::Duration::zero() {
+            bail!("cannot convert negative interval to duration");
+        }
+        Ok(self.duration_as_chrono().to_std()?)
```
What does chrono buy us here as we're immediately throwing it away? The negative detection?

jkosh44 replied:
It was purely for convenience to try and avoid code duplication. I've removed it.

mjibson commented on src/repr/src/adt/interval.rs:
```diff
@@ -388,6 +435,99 @@ impl Interval {
+        }
+    }
+
+    fn convert_date_time_unit_increasing<T>(
```
Ooh this is a neat function.

mjibson reviewed (commented):
This is looking better. I think it's worth figuring out the PartialOrd, Ord, Eq stuff. I thought that we thought we could get ord working but not eq when we discussed this last week?

sploiselle commented on src/expr/src/scalar/func/impls/time.rs:
```diff
@@ -28,11 +28,12 @@ sqlfunc!(
-        )
-        .map_err(|_| EvalError::IntervalOutOfRange)
+        // wont overflow because value can't exceed 24 hrs + 1_000_000 ns = 86_400 seconds + 1_000_000 ns = 86_400_001_000 us
+        let micros: i64 = i64::from(t.num_seconds_from_midnight())
```
Could this use `convert_date_time_unit`?

jkosh44 replied:
I changed it to use `convert_date_time_unit` for the second to microsecond conversion, but that function doesn't support nanosecond so that stayed manual.

sploiselle commented on src/repr/src/adt/datetime.rs:
```diff
@@ -201,49 +202,21 @@ impl DateTimeField {
-    /// Panics if called on a non-duration field or a field larger than or equal to a second.
-    pub fn seconds_divider(self) -> i64 {
+    /// Panics if called on a non-time/day field.
+    pub fn micros_multiplier(self) -> i64 {
```
Seems like this wants to either be powered by or replaced by `convert_date_time_unit`

jkosh44 replied:
Good point, I've changed the implementation to use `convert_date_time_unit`

sploiselle commented on src/repr/src/adt/datetime.rs:
```diff
@@ -349,6 +322,10 @@ impl DateTimeFieldValue {
+
+    /// How much padding is added to the fractional portion to achieve a given precision.
+    /// e.g. with the current precision `.5` is represented as `500_000_000`.
+    const FRACTIONAL_DIGIT_PRECISION: i64 = 1_000_000_000;
```
Could this be replaced by `convert_date_time_unit`?

jkosh44 replied:
To be honest, I'm not exactly sure how to go about that. This is used to convert between a decimal part of a float and an integer. For example 0.123456789 to 1234556789. It's not exactly about converting between different time/date units.

sploiselle reviewed (commented):
No good deed goes unpunished! Really great work cleaning a lot of this up, but have questions about a few spots I think could get folded into `convert_date_time_unit`

jkosh44 commented:
> What's the minimum amount of chrono we need for this? I think using chrono should be used only when we are forced to because an interval has to interact with some other type that's backed by chrono, but otherwise it should avoid chrono completely.

@mjibson I agree that removing chrono is probably a good idea. However, I'm a bit hesitant to remove it because this is already such a big PR without that change, it might be best left to a follow up PR.

mjibson commented:
Followup SGTM.

jkosh44 commented on src/repr/src/adt/datetime.rs:
```diff
@@ -201,50 +202,24 @@ impl DateTimeField {
-            Milliseconds => 1_000,
-            Microseconds => 1_000_000,
-            _other => unreachable!("Do not call with a field larger than or equal to a second"),
+            Day => {}
```
Fixed

jkosh44 commented on src/repr/src/adt/datetime.rs:
```diff
@@ -255,12 +230,14 @@ impl DateTimeField {
-            Century => 12 * 100,
-            Decade => 12 * 10,
-            Year => 12,
+            Millennium => {}
```
Fixed

mjibson reviewed (approved):
Wow this is great. In some followup work I'm very interested in improving the ordering semantics or describing why they can't be done due to 10566. I thought that issue was mostly about equality for our current types, and I'd like to learn more if that's wrong.

jkosh44 commented:
> This is looking better. I think it's worth figuring out the PartialOrd, Ord, Eq stuff. I thought that we thought we could get ord working but not eq when we discussed this last week?

@mjibson @sploiselle 

My understanding is that @benesch was saying that there's no requirement in Dataflow/Materialize for ordering to be deterministic, that limitation only exists with equality. So as far as Dataflow/Materialize is concerned, we can implement `PartialOrd` and `Ord` on `Interval` such that we consider 24 hours equal to 1 day (and 30 days equal to 1 month).

However I do think there is a limitation in Rust that prevents this. When I went to implement `PartialOrd` I noticed this in the documentation (https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html)
> The methods of this trait must be consistent with each other and with those of PartialEq in the following sense:
     a == b if and only if partial_cmp(a, b) == Some(Equal).
     a < b if and only if partial_cmp(a, b) == Some(Less) (ensured by the default implementation).
     a > b if and only if partial_cmp(a, b) == Some(Greater) (ensured by the default implementation).
     a <= b if and only if a < b || a == b (ensured by the default implementation).
     a >= b if and only if a > b || a == b (ensured by the default implementation).
     a != b if and only if !(a == b) (already part of PartialEq).

So I'm not sure that's it's possible to implement `PartialOrd` in a "sane" way where we consider 24 hours == 1 day without also implementing `Eq` and `PartialEq` to consider 24 hours == 1 day. Though I'm interested in hearing other people's thoughts, I may be misunderstanding the `PartialOrd` requirement.

One workaround I was considering is that we arbitrarily say that 1 day > 24 hours, and 1 day < 24 hours 1 microsecond. That may allow us to sanely order Interval's without breaking the requirement Rust imposes on `PartialOrd` and `PartialEq`.

jkosh44 commented:
@sploiselle @mjibson I believe I've addressed all of the comments left on the PR. If I missed something, please let me know. I'm going to focus on getting it through CI now unless there's more comments or something I missed.

mjibson commented:
> we arbitrarily say that 1 day > 24 hours, and 1 day < 24 hours 1 microsecond

Yes, this seems like what we should do. I think that means everything is correct except for the edge cases of (24h, 1d) and (30d, 1month), and 10566 will fix those.

jkosh44 commented:
> > we arbitrarily say that 1 day > 24 hours, and 1 day < 24 hours 1 microsecond
> 
> Yes, this seems like what we should do. I think that means everything is correct except for the edge cases of (24h, 1d) and (30d, 1month), and 10566 will fix those.

Ok, I think that makes sense. Though I still think it's probably best left for a follow-up issue. It's still a bit tricky considering some different cases.

- 1 day > 24 hrs
- 2 days > 48 hrs 
- x months 1 day > x months 24 hrs
- x months 1 day ?? y months 24 hours

So we won't be able to just special case a couple of scenarios and will have to handle the logic gracefully.

mjibson commented:
Yes, followup. I concur with your decision to get this in and just fix CI failures now.
